invenio_app_ils/holidays/api.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# invenio-app-ils is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""ILS holidays API."""
+
+from datetime import timedelta
+
+import arrow
+
+from invenio_app_ils.proxies import current_app_ils
+
+_day_increment = timedelta(days=1)  # Atomic increment
+
+
+def _is_normally_open(location, date):
+    """Checks if the location is normally opened on a given date wrt to the regular schedule."""
+    opening = location["opening_weekdays"]
+    weekday_number = date.weekday()
+    opening_weekday = opening[weekday_number]
+    return opening_weekday["is_open"]
+
+
+def _is_exceptionally_open(location, date):
+    """Checks if the location is exceptionally opened on a given date, or None if undefined."""
+    for exception in location["opening_exceptions"]:
+        if date <= arrow.get(exception["end_date"]).date():
+            if arrow.get(exception["start_date"]).date() <= date:
+                return exception["is_open"]
+            else:
+                return None  # Early return
+    return None  # Exhaustion
+
+
+def is_open_on(location_pid, date):
+    """Checks if the location is open or closed on a given date."""
+    location = current_app_ils.location_record_cls.get_record_by_pid(location_pid)
+    exceptionally_open = _is_exceptionally_open(location, date)
+    if exceptionally_open is not None:
+        return exceptionally_open
+    return _is_normally_open(location, date)
+
+
+def next_open_after(location_pid, date):
+    """Finds the next day where this location is open."""
+    location = current_app_ils.location_record_cls.get_record_by_pid(location_pid)
+    n_weekdays = len(location["opening_weekdays"])
+    weekdays_closed = False
+    for exception in location["opening_exceptions"]:
```

Comment:
lets say that `date="2020-09-10"`
our next opened day id `2020-09-11`
as I understood it, it will still go into this loop and check all the exceptions even though the next day does not belong to an exception, correct me if I am wrong

Reply:
You're totally right, that's why we decided to remove the exceptions that expired with a cron that is run everyday (currently what Javier is working on). Therefore in practice there will only be one (or more rarely a few more) iterations in the loop.
Otherwise we can decide to move to a dedicated schema to store the exceptions and then create a scanner that will iterate over all the exceptions that end _after today_. The complexity would be similar.